### PR TITLE
feat(minifer): minify number

### PR DESCRIPTION
**Pull request by mysteryven**
Port [printNonNegativeFloat](https://github.com/evanw/esbuild/blob/e6a8169c3a574f4c67d4cdd5f31a938b53eb7421/internal/js_printer/js_printer.go#L3225) from ESBuild.

### CODE REVIEW

File: tasks/minsize/minsize.snap [hunk 1]
```diff
@@ -1,25 +1,25 @@
 Original   -> Minified   -> Gzip       Brotli    
 72.14 kB   -> 24.36 kB   -> 8.68 kB    7.64 kB    react.development.js
 
-173.90 kB  -> 61.97 kB   -> 19.46 kB   17.76 kB   moment.js
+173.90 kB  -> 61.89 kB   -> 19.47 kB   17.70 kB   moment.js
```
Comment (mysteryven) [inline tasks/minsize/minsize.snap:4]: Why the Gzip becomes larger? I don't really know.

File: crates/oxc_minifier/src/printer/gen.rs [hunk 2]
```diff
@@ -793,23 +793,88 @@ impl Gen for NullLiteral {
...
+
+        // We'll need a space before "." if it could be parsed as a decimal point
+        if !bytes.iter().any(|&b| matches!(b, b'.' | b'e' | b'x')) {
+            p.need_space_before_dot = p.code_len();
         }
-        p.print_str(self.raw.as_bytes());
     }
 }
 
+fn print_non_negative_float(value: f64, p: &mut Printer) -> String {
```
Comment (Boshen) [inline crates/oxc_minifier/src/printer/gen.rs:824]: The algorithm works, but I think we should be able to refactor this into the least number of allocations. A small task @anonrig may be able to help in the future.

File: tasks/minsize/minsize.snap (hunk 1 above)
Comment (Boshen) [reply tasks/minsize/minsize.snap:4]: This is fine, we want the overall size to go smaller first.

File: crates/oxc_minifier/src/printer/gen.rs [hunk 3]
```diff
@@ -879,8 +944,7 @@ impl<'a> GenExpr for StaticMemberExpression<'a> {
         self.object.gen_expr(p, Precedence::Postfix);
         if self.optional {
             p.print(b'?');
-        } else if matches!(&self.object, Expression::NumberLiteral(n) if n.base == NumberBase::Decimal)
-        {
+        } else if p.need_space_before_dot == p.code_len() {
```
Comment (Boshen) [inline crates/oxc_minifier/src/printer/gen.rs:947]: NICE!! This was the line I was missing, thank you for catching this

File: crates/oxc_minifier/src/printer/gen.rs (hunk 2 above)
Comment (Boshen) [reply crates/oxc_minifier/src/printer/gen.rs:824]: Can you add a `TODO: refactor this with less allocations` comment?

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Let me know once this is mergeable.

**Comment by mysteryven** [comment]
> Let me know once this is mergeable.

Done, I only add a comment :)

**Comment by Boshen** [comment]
Thank you so much for working on this.

**Pull request merged**